src/faiss.cc

```diff
@@ -5,80 +5,105 @@
-      try
-      {
-        index_ = std::unique_ptr<faiss::IndexFlatL2>(dynamic_cast<faiss::IndexFlatL2 *>(faiss::read_index(fname.c_str())));
-      }
-      catch (const faiss::FaissException& ex)
-      {
-        Napi::Error::New(env, ex.what()).ThrowAsJavaScriptException();
-      }
+      Napi::Error::New(env, "Class constructors cannot be invoked without 'new'")
+          .ThrowAsJavaScriptException();
+      return;
     }
-    else
+    if (info.Length() > 0 && info[0].IsNumber())
     {
-      if (!info.IsConstructCall())
-      {
-        Napi::Error::New(env, "Class constructors cannot be invoked without 'new'")
-            .ThrowAsJavaScriptException();
-        return;
-      }
-
-      if (info.Length() != 1)
-      {
-        Napi::Error::New(env, "Expected 1 argument, but got " + std::to_string(info.Length()) + ".")
-            .ThrowAsJavaScriptException();
-        return;
-      }
-      if (!info[0].IsNumber())
-      {
-        Napi::TypeError::New(env, "Invalid the first argument type, must be a number.").ThrowAsJavaScriptException();
-        return;
-      }
-
       auto n = info[0].As<Napi::Number>().Uint32Value();
-      index_ = std::unique_ptr<faiss::IndexFlatL2>(new faiss::IndexFlatL2(n));
+      index_ = std::unique_ptr<Y>(new Y(n));
     }
   }
 
-  static Napi::Object Init(Napi::Env env, Napi::Object exports)
+  static Napi::Value read(const Napi::CallbackInfo &info)
   {
-    // clang-format off
-    Napi::Function func = DefineClass(env, "IndexFlatL2", {
-      InstanceMethod("ntotal", &IndexFlatL2::ntotal),
-      InstanceMethod("getDimension", &IndexFlatL2::getDimension),
-      InstanceMethod("isTrained", &IndexFlatL2::isTrained),
-      InstanceMethod("add", &IndexFlatL2::add),
-      InstanceMethod("search", &IndexFlatL2::search),
-      InstanceMethod("write", &IndexFlatL2::write),
-      InstanceMethod("mergeFrom", &IndexFlatL2::mergeFrom),
-      InstanceMethod("removeIds", &IndexFlatL2::removeIds),
-      StaticMethod("read", &IndexFlatL2::read),
-    });
-    // clang-format on
+    Napi::Env env = info.Env();
 
-    Napi::FunctionReference *constructor = new Napi::FunctionReference();
-    *constructor = Napi::Persistent(func);
-    env.SetInstanceData(constructor);
+    if (info.Length() != 1)
+    {
+      Napi::Error::New(env, "Expected 1 argument, but got " + std::to_string(info.Length()) + ".")
+          .ThrowAsJavaScriptException();
+      return env.Undefined();
+    }
+    if (!info[0].IsString())
+    {
+      Napi::TypeError::New(env, "Invalid the first argument type, must be a string.").ThrowAsJavaScriptException();
+      return env.Undefined();
+    }
 
-    exports.Set("IndexFlatL2", func);
-    return exports;
+    Napi::Object instance = T::NewInstance(env, {});
+    T *index = Napi::ObjectWrap<T>::Unwrap(instance);
+
+    std::string fname;
+    if (info[0].IsExternal())
+    {
+      fname = *info[0].As<Napi::External<std::string>>().Data();
```

Comment:
It seems that this branch will not reach. maybe we can get `fname` directly:
`std::string fname = info[0].As<Napi::String>().Utf8Value();`